Add test for HTML compliance: <marquee> must use the HTMLMarqueeElement interface

tomhodgins opened the pull request:
Hello, I noticed that Servo doesn't follow the HTML spec when it comes to the requirements surrounding how it handles `<marquee>` elements so I created this PR to test [the HTML spec's](https://html.spec.whatwg.org/multipage/obsolete.html#the-marquee-element) assertion that:

> The marquee element must implement the `HTMLMarqueeElement` interface.

This is my first web platform test, so please let me know if there's anything I need to do to package this up differently.

Thanks!

tomhodgins committed:
Add test for HTML compliance: <marquee> must use the HTMLMarqueeElement interface

domenic commented:
So, we already have tests for every element interface <-> name combination, but in kind of a strange place: https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/Node-cloneNode.html.

However that failed to catch this because https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/Node-cloneNode.html#L123 maps `marquee` to `HTMLElement`, not to `HTMLMarqueeElement`. So I suggest changing that line, instead of introducing a new file.

Additionally, it might be best to change https://github.com/web-platform-tests/wpt/blob/master/dom/nodes/Node-cloneNode.html#L39-L40 to use `.constructor.name === type.name` instead of `instanceof type`. That would avoid this sort of problem in the future. (And might uncover some other mismappings in that file.)

tomhodgins commented:
Hi @domenic, thanks for the input. If I update lines 39-40 to use `.constructor.name === type.name` it does reveal some additional mis-mappings, and also a few of the other tests break as well relating to documents and cloning elements.

> line 65: create_element_and_check("bgsound",   "HTMLElement");

I get `HTMLUnknownElement` in Chrome

> line 67: create_element_and_check("blockquote","HTMLElement");

I get `HTMLQuoteElement` in Chrome

> line 83: create_element_and_check("details",   "HTMLElement");

I get `HTMLDetailsElement` in Chrome

> line 114: create_element_and_check("isindex",   "HTMLElement");

I get `HTMLUnknownElement` in Chrome

> line 150: create_element_and_check("spacer",    "HTMLElement");

I get `HTMLUnknownElement` in Chrome

And some other browsers don't agree about `<dialog>`, etc.

What's the best thing to do here?

domenic commented:
>    line 65: create_element_and_check("bgsound", "HTMLElement");
>
> I get HTMLUnknownElement in Chrome

Chrome matches the spec, so updating the file would be good.

>    line 67: create_element_and_check("blockquote","HTMLElement");
>
>I get HTMLQuoteElement in Chrome

Chrome matches the spec, so updating the file would be good.

>    line 83: create_element_and_check("details", "HTMLElement");
>
> I get HTMLDetailsElement in Chrome

Chrome matches the spec, so updating the file would be good.

>    line 114: create_element_and_check("isindex", "HTMLElement");
>
> I get HTMLUnknownElement in Chrome

Chrome matches the spec, so updating the file would be good.

>    line 150: create_element_and_check("spacer", "HTMLElement");
>
> I get HTMLUnknownElement in Chrome

Chrome matches the spec, so updating the file would be good.

> And some other browsers don't agree about <dialog>, etc.

Chrome matches the spec, so _not_ updating the file would be good :).

tomhodgins committed:
Delete marquee-element-interface.html, Update Node-cloneNode.html to compare element constructor names in lieu of using instanceof to compare type

tomhodgins commented:
I've removed the new file I had created, and updated the `dom/nodes/Node-cloneNode.html` file

- I changed the element comparisons from `instanceof` to `constructor.name`
- I updated the expected HTML element names to conform to the spec
- I created a function named `check_document_copy()` which performs the same `instanceof` comparisons that can continue to be used for the tests comparing document objects

Were these the right calls to make here? Thanks for the suggestions!

annevk commented on dom/nodes/Node-cloneNode.html:
```diff
@@ -34,6 +34,13 @@
 function check_copy(orig, copy, type) {
+    assert_not_equals(orig, copy, "Object equality");
+    assert_equal_node(orig, copy, "Node equality");
+    assert_true(orig.constructor.name === type.name, "original instanceof " + type);
+    assert_true(copy.constructor.name === type.name, "copy instanceof " + type);
```
If you use assert_equals here you'll get more helpful error messages and the message argument can basically be "original" and "copy", respectively.

annevk reviewed (commented):
We should probably use constructor.name for the document things too, but I suspect browsers disagree a fair bit there still.

domenic commented on dom/nodes/Node-cloneNode.html:
```diff
@@ -198,19 +205,19 @@
   el.setAttribute("c", "d");
   var c = el.cloneNode();
-  check_copy(el, c, Element);
+  check_copy(el, c, {name: "HTMLUnknownElement"});
```
```suggestion
  check_copy(el, c, HTMLUnknownElement);
```

would work, I think? Same for the other ones; I don't think they needed to be changed?

domenic reviewed (commented):
Generally looks great, with a nit and @annevk's nit as well.
